### PR TITLE
Make TiktokenTokenizerWrapper picklable

### PR DESCRIPTION
**Pull request by irenedea**
Makes TiktokenTokenizerWrapper serializable so we can use it with multi-processing.

# Tests
1. Finetune run with tiktoken tokenizer: test-pickle-tiktoken-ohOd0p
2. Works with packing
3.  pickle code:
```
from llmfoundry.tokenizers import TiktokenTokenizerWrapper
t = TiktokenTokenizerWrapper('gpt2')
import pickle
pickle.dumps(t) # works
```
(Previously, the above code would raise an error: `TypeError: cannot pickle 'builtins.CoreBPE' object`)

### CODE REVIEW

File: llmfoundry/tokenizers/tiktoken.py [hunk 1]
```diff
@@ -50,6 +50,21 @@ def __init__(self,
...
+        import functools
+
+        def pickle_Encoding(enc: tiktoken.Encoding):
+            return (functools.partial(tiktoken.Encoding,
+                                      enc.name,
+                                      pat_str=enc._pat_str,
+                                      mergeable_ranks=enc._mergeable_ranks,
+                                      special_tokens=enc._special_tokens), ())
+
+        copyreg.pickle(tiktoken.core.Encoding, pickle_Encoding)
```
Comment (irenedea) [inline llmfoundry/tokenizers/tiktoken.py:66]:
```suggestion
        copyreg.pickle(tiktoken.Encoding, pickle_Encoding)
```

**Comment by samhavens** [comment]
Do you know why we are getting ` /home/runner/work/llm-foundry/llm-foundry/build/lib/llmfoundry/tokenizers/tiktoken.py:59:43 - error: "Encoding" is not a known member of module (reportGeneralTypeIssues)` in the Code Quality checks? I was getting this too. Is the CI environment not getting a full install of the tokenizer or something?

**Comment by irenedea** [comment]
@samhavens I think we might need to add the openai deps to the .github workflow file. Will try it!

File: .github/workflows/code-quality.yaml [hunk 2]
```diff
@@ -27,7 +27,7 @@ jobs:
         - '3.9'
         - '3.10'
         pip_deps:
-        - '[dev]'
+        - '[dev,openai]'
```
Comment (irenedea) [inline .github/workflows/code-quality.yaml:30]: gonna remove and use type ignores

File: llmfoundry/tokenizers/tiktoken.py [hunk 3]
```diff
@@ -50,6 +50,23 @@ def __init__(self,
             raise ImportError(
                 'You need to install tiktoken to use TiktokenTokenizerWrapper.')
 
+        # Workaround to make tiktokenizer picklable.
+        # https://github.com/huggingface/datasets/issues/5536#issuecomment-1682309347
+        # There is an open PR from HF to add this to tiktoken: https://github.com/openai/tiktoken/pull/181
+        import copyreg
+        import functools
+
+        from tiktoken import Encoding  # type: ignore (thirdParty)
```
Comment (irenedea) [inline llmfoundry/tokenizers/tiktoken.py:59]: doing tiktoken.Encoding does not work and I cannot find a way to satisfy pyright that way, so I changed this to import Encoding directly. I think it's related to this (https://github.com/microsoft/pyright/issues/4863 ), but I don't totally understand what pyright can model and cannot model w.r.t. to imports

Comment (dakinggg) [reply llmfoundry/tokenizers/tiktoken.py:59]: type ignore on the line that used tiktoken.Encoding did not work?

Comment (irenedea) [reply llmfoundry/tokenizers/tiktoken.py:59]: nope, I tried that, but it was erroring on a line that did not even use tiktoken.Encoding, and even adding type ignore to that line didn't work...

Comment (irenedea) [reply llmfoundry/tokenizers/tiktoken.py:59]: https://github.com/mosaicml/llm-foundry/actions/runs/6699130750/job/18202680494 this was the failure:
`home/runner/work/llm-foundry/llm-foundry/llmfoundry/tokenizers/tiktoken.py:70:33 - error: "Encoding" is not a known member of module`, L70 doesn't even contain any ref to tiktoken.Encoding and adding a type ignore there didn't work

Comment (dakinggg) [reply llmfoundry/tokenizers/tiktoken.py:59]: ok, sure 🤷‍♂️ im not gonna think about this too hard. whatever works

Comment (irenedea) [reply llmfoundry/tokenizers/tiktoken.py:59]: haha yeah, well this way we only need one type ignore also 😄

**Review by dakinggg — APPROVED** [review]

**Pull request merged**